homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,200 @@
+                reconnect_callback=reconnected,
+                loop=hass.loop,
+                logger=_LOGGER,
+                timeout=CONNECTION_TIMEOUT,
+                reconnect_interval=DEFAULT_RECONNECT_INTERVAL)
+
+            hass.data[DATA_DEVICE_REGISTER][device] = client
+
+            # Load platforms
+            for comp_name, comp_conf in switches.items():
+                hass.async_create_task(
+                    async_load_platform(hass, 'switch', DOMAIN,
+                                        (comp_name, comp_conf, device),
+                                        config))
+
+            # handle shutdown of HLK-SW16 asyncio transport
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                       lambda x: client.stop())
+
+            _LOGGER.info('Connected to HLK-SW16')
+
+        hass.loop.create_task(connect())
+
+    for device in config[DOMAIN]:
+        add_device(device)
+    return True
+
+
+class SW16Device(Entity):
+    """Representation of a HLK-SW16 device.
+
+    Contains the common logic for HLK-SW16 entities.
+    """
+
+    def __init__(self, relay_name, device_port, device_id, client):
+        """Initialize the device."""
+        # HLK-SW16 specific attributes for every component type
+        self._device_id = device_id
+        self._device_port = device_port
+        self._is_on = None
+        self._client = client
+        self._name = relay_name
+
+    @callback
+    def handle_event_callback(self, event):
+        """Propagate changes through ha."""
+        _LOGGER.debug("Relay %s new state callback: %r",
+                      self._device_port, event)
+        self._is_on = event
+        self.async_schedule_update_ha_state()
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return a name for the device."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._is_on
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return bool(self._client.is_connected)
+
+    @callback
+    def _availability_callback(self, availability):
+        """Update availability state."""
+        self.async_schedule_update_ha_state()
+
+    async def async_added_to_hass(self):
+        """Register update callback."""
+        self._client.register_status_callback(self.handle_event_callback,
+                                              self._device_port)
+        self._is_on = await self._client.status(self._device_port)
+        async_dispatcher_connect(self.hass, SIGNAL_AVAILABILITY,
+                                 self._availability_callback)
+
+
+class SwitchableSW16Device(SW16Device):
+    """HLK-SW16 entity which can switch on/off (eg: light, switch)."""
+
+    async def async_update(self):
```

Comment:
This will never be called since the entity is non polling and we don't call this method explicitly.